test/spec/provider/cloud-element-templates/cmd/ChangeElementTemplateHandler.spec.js

```diff
@@ -640,6 +640,69 @@ describe('cloud-element-templates - ChangeElementTemplateHandler', function() {
 
     });
 
+
+    describe('update task type', function() {
+
+      const newTemplate = require('./task-template-elementType.json');
+
+      beforeEach(bootstrap(require('./task.bpmn').default, [ newTemplate ]));
+
+      it('execute', inject(function(elementRegistry) {
+
+        // given
+        const task = elementRegistry.get('Task_1');
+
+        // assume
+        expect(is(task, 'bpmn:ServiceTask')).to.be.true;
+
+        // when
+        changeTemplate(task, newTemplate);
+
+        // then
+        const replacedTask = elementRegistry.get('Task_1');
```

Comment:
Do we need to access it twice? It should also work if we check the type for `task`, right?

Reply:
As the task get's replaced during `bpmnReplace.replaceElement`, the reference of `task` is outdated. 

To consider: what is the return value of `commandStack.execute('propertiesPanel.zeebe.changeTemplate')`? Currently it is undefined, but maybe we can do something like `{template, element}` with the replace element + new template. WDYT?

Reply:
I have a bit of feeling that this could cause troubles on consumers side. Currently, when using the command, I don't have to care about the return value, I just can rely that my provided `element` will be updated.

> the reference of `task` is outdated

Is that really the case? Isn't [this line](https://github.com/bpmn-io/bpmn-js-properties-panel/pull/648#discussion_r839580439) preventing this?

Reply:
> Is that really the case? Isn't https://github.com/bpmn-io/bpmn-js-properties-panel/pull/648#discussion_r839580439 preventing this?

Unfortunately not. It just updates the reference for the method so the attributes are applied to the new element. All references that were gotten externally (e.g. from the `elementRegistry`) still hold the old element that got replaced

Reply:
> Currently, when using the command, I don't have to care about the return value, I just can rely that my provided element will be updated.

I'm not sure if we can still support this if we have to change element types. If we can't, this would be a breaking change and we need a clear API on how to get a current reference to the element

Reply:
I think this is something we need to clarify carefully /cc @marstamm @nikku

Reply:
For me this would be intended and also requested as outlined [here](https://github.com/bpmn-io/bpmn-js-properties-panel/issues/642). I understand it as: 

* In the past, we primarily applied element templates on top of existing element templates. So a `changeTemplate` operation would not change the element per so, but just what is applied on top of it.
* Now conceptually, we see an element template as a "custom" task. So when you apply an template, you conceptually really make it a different kind of task. So a `changeTemplate` would actually change the element per se.
* (There are edge cases like when the template has still the same targetType and so on, but what I wrote above, would be the basic conceptual architecture)

So I would say, yes, this breaks existing behavior, but this is intended and okay.

Reply:
I agree that this behavior might be expected now. However, [we have API](https://github.com/bpmn-io/bpmn-js-properties-panel/blob/master/src/provider/element-templates/ElementTemplates.js#L133) in place that returns the element. This is public API, so we need to adapt and communicate the breaking change.

Since the API was added in alpha, I don't expect negative consequences here. We just to have to make sure we don't rely on it in our own implementations where we use the command.

* https://github.com/camunda/camunda-modeler/blob/develop/client/src/plugins/element-templates-modal/modeler/EditorActions.js#L19
* https://github.com/bpmn-io/element-template-chooser/blob/master/src/element-template-chooser/ElementTemplateChooser.js#L75

Reply:
> I have a bit of feeling that this could cause troubles on consumers side. Currently, when using the command, I don't have to care about the return value, I just can rely that my provided element will be updated.

At the moment whatever our users do is fiddling with internals. We've just now created actual APIs to rely on, via https://github.com/bpmn-io/bpmn-js-properties-panel/pull/624.

And the new API addition is built with the replace use-case in mind already.

Reply:
Thank you for the insights.

> However, [we have API](https://github.com/bpmn-io/bpmn-js-properties-panel/blob/master/src/provider/element-templates/ElementTemplates.js#L133) in place that returns the element.

Would this be a breaking change if we return the new element instead?

Reply:
Last comment: Until this point element templates was truely "experimental", and also marked as such, everywhere (cf. [docs](https://docs.camunda.io/docs/components/modeler/desktop-modeler/element-templates/about-templates/)).

With 1.0 we're getting closer with actually _stabilizing it_. The feature proves itself holisticlly, not only hidden in the properties panel.